TensroIterator preserve format for binary, ternary operators.

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29461 [WIP] Experiments
* #29460 [WIP] Switch default of clone, to, *_like ops
* #29459 [WIP] Prepare copy op to switch memory format preservation defaults
* #28991 Adding function to convert Module to channels last
* #28544 Add torch.memory_format support to the TorchScript
* #28292 Add memory format support to the `resize_` op.
* **#28291 TensroIterator preserve format for binary, ternary operators.**
* #28077 Add channels last support to cuda.comm.scatter and gather
* #28076 Kill `operator==` of TensorOptions as confusing one
* #27979 Add memory format support to `resize_as_` operator

Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

VitalyFedyunin committed:
TensroIterator preserve format for binary, ternary operators.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "TensroIterator preserve format for binary, ternary operators."




[ghstack-poisoned]

VitalyFedyunin committed:
Fix rand_like test on "TensroIterator preserve format for binary, ternary operators."




[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_as_ linter on "TensroIterator preserve format for binary, ternary operators."




[ghstack-poisoned]

VitalyFedyunin committed:
Fix onnx on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
Rebase on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
randn_like not simple mapable any more on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
randn_like mappable again and supported in jit with memory format on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
fix bc test for empty_like on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
fix bc test for empty_like now with gradient on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin commented:
@ngimel I will add tests and benchmark results, so far I just looking your opinion on structure.

VitalyFedyunin committed:
rebase to master on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
add TensroIterator tests on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
Add some tests on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
Addressed PR feedback on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
Linter fix on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

ngimel commented on test/test_torch.py:
```diff
@@ -11956,6 +11957,88 @@ def test_memory_format_empty_like(self, device):
+            # lambda x, y: x.abs(), # https://github.com/pytorch/pytorch/issues/24531
+            # lambda x, y: x.acos(), # https://github.com/pytorch/pytorch/issues/24532
+            lambda x, y: x.add(y, alpha=3),
+            lambda x, y: x.addcdiv(2, y, y),
```
given that y is contiguous, what are we testing here? Should it have been x, y in the arguments (same for addcmul)

VitalyFedyunin replied:
Addressed by adding more tests

ngimel commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -275,33 +275,51 @@ DimVector TensorIterator::invert_perm(IntArrayRef input) const {
         op.tensor = at::empty(tensor_shape, op.options());
+        op.tensor.unsafeGetTensorImpl()->empty_tensor_restride(
+            MemoryFormat::ChannelsLast);
+        op.stride_bytes = apply_perm_and_mul(op.tensor.strides(), element_size);
```
I cannot convince myself that `op.stride_bytes` and `op.tensor.strides()` are guaranteed to be consistent (one is defined by perm_ which can be whatever, and the other is hardcoded nhwc), and tests are only testing contiguity of output which is guaranteed by `empty_tensor_restride`, not its correctness.

VitalyFedyunin replied:
Addressed by adding comment

VitalyFedyunin committed:
Fix TensorIterator on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix Autograd on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_as_ with new test on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_as_ on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
Add resize_as_ docs on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
fix resize_as_ docs on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]

VitalyFedyunin committed:
fix resize_as_ docs on "TensroIterator preserve format for binary, ternary operators."


Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

[ghstack-poisoned]
